target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8074-x80-5g.dts

```diff
@@ -0,0 +1,476 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/* Copyright (c) 2022, Robert Marko <robimarko@gmail.com> */
+/* altered by Jonathan Brophy <professor_jonny@hotmail.com> */
+
+/dts-v1/;
+
+#include "ipq8074.dtsi"
+#include "ipq8074-hk-cpu.dtsi"
+#include "ipq8074-ess.dtsi"
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+#include <dt-bindings/leds/common.h>
+
+/ {
+	model = "tplink x80-5g";
+	compatible = "tplink,x80-5g", "qcom,ipq8074";
+
+	aliases {
+		 //Aliases as required by u-boot to patch MAC addresses
+		 serial0 = &blsp1_uart5;
+ 		 ethernet0 = &dp5; // Aliases to capture wan mac from bootloader
+	};
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+		bootargs-append = " root=/dev/ubiblock0_1";
+	};
+
+	modem {
+		compatible = "gpio-export";
+		pinctrl-0 = <&modem_pins>;
+		pinctrl-names = "default";
+
+		w_disable {
+			gpio-export,name = "w_disable";
+			gpio-export,output = <1>;
+			gpios = <&tlmm 35 GPIO_ACTIVE_HIGH>; // wwan disable output
+		};
+
+		modem-reset {
+			gpio-export,name = "modem_reset";
+			gpio-export,output = <0>;
+			gpios = <&tlmm 29 GPIO_ACTIVE_HIGH>; // reset modem output
+		};
+	};
+
+		antennas {
+		compatible = "gpio-export";
+		pinctrl-0 = <&antenna_pins>;
+		pinctrl-names = "default";
+
+		ext_antenna_en {
+			gpio-export,name = "ext_antenna_en";
+			gpio-export,output = <0>;
+			gpios = <&tlmm 56 GPIO_ACTIVE_HIGH>; // select external sma antennas for WWan output
+		};
+	};
+
+	gpio_keys {
+		compatible = "gpio-keys";
+		pinctrl-0 = <&button_pins>;
+		pinctrl-names = "default";
+
+		button@1 {
+			label = "reset";
+			gpios = <&tlmm 27 GPIO_ACTIVE_LOW>; //reset button output
+			linux,code = <KEY_RESTART>; // Aliases for button to stitch to linux reset functioon
+		};
+	};
+
+	leds {
+		compatible = "gpio-leds";
+		pinctrl-0 = <&led_pins>;
+		pinctrl-names = "default";
+		// note the LED'd defined here are not individual they are a combined RGB led
+		// and each led has no defined single function but has multiple functioons
+		//to mimic the oem default behaviour green and red are switched on from boot to indicate yellow
+		//to show the device is booting behaviour can be user defined in procd to override this.
+
+		led_system_green: green {
+			gpios = <&tlmm 50 GPIO_ACTIVE_HIGH>; //green led component of RGB led
+			color = <LED_COLOR_ID_GREEN>;
+			default-state = "on";
+		};
+
+		led_system_red: red {
+			gpios = <&tlmm 51 GPIO_ACTIVE_HIGH>; //red led component of RGB led
+			color = <LED_COLOR_ID_RED>;
+			default-state = "on";
+		};
+
+		led_system_blue: blue {
+			gpios = <&tlmm 52 GPIO_ACTIVE_HIGH>; //blue led component of RGB
+			color = <LED_COLOR_ID_BLUE>;
+		};
```

Comment:
Please add missing `function` properties.

Reply:
And please use `led-*` aliases in `aliases {}` node for status LED handling of OpenWrt instead of user-space script.

Reply:
The LED's don't have a set assigned function they have multiple as they are not singular LED's it is a multifunction RGB LED.
There is no monochromatic LED support in OpenWrt and no support for virtual LED's' by mixing colors to be able to assign individual functions to individual colors.
I don't know how to proceed, and other targets has left them blank as I have.

Reply:
`function` property is used for LED name `<color>:<function>` (example: `green:status`) in **/sys/class/leds/**.

IMO:

```
aliases {
	led-boot = &led_system_green;
	led-falsafe = &led_system_red;
	led-running = &led_system_green;
	led-upgrade = &led_system_green;
	...
}
...
led_system_green: led-0 {
	gpios = <&tlmm 50 GPIO_ACTIVE_HIGH>;
	color = <LED_COLOR_ID_GREEN>;
	function = LED_FUNCTION_STATUS;
};

led_system_red: led-1 {
	gpios = <&tlmm 51 GPIO_ACTIVE_HIGH>;
	color = <LED_COLOR_ID_RED>;
	function = LED_FUNCTION_STATUS;
	default-state = "off";
};

led-2 {
	gpios = <&tlmm 52 GPIO_ACTIVE_HIGH>;
	color = <LED_COLOR_ID_BLUE>;
	function = LED_FUNCTION_STATUS;
	default-state = "off";
};
```

Reply:
how do I assign an alias to mimic the oem colours ? as power is yellow in the device which is a mix of red and green led's

As at the moment I have left the default state as on for both of those led's as it seems I can only have a one to one mapping as there is no virtual gpio function like the QSDK.

`
		yellow_mix {
			default-state = "off";
			virtuals = <red_led>, <green_led>;
		};
		blue_mix {
			default-state = "off";
			virtuals = <blue_led>, <green_led>;

Reply:
If we could implement this then I can define LED colours direcly in the dts I believe.

 https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/drivers/leds/rgb/leds-group-multicolor.c?h=v6.10.8

Reply:
I have just removed the script and will drive from procd, and I plan to make a kmod for the led's to support colour mixing the led's directly from DTS.